### DISCUSSION TITLE
JSON column support for oracledb Oracle dialect

### DISCUSSION THREAD

**Discussion by TanawanK** [sqlalchemy/sqlalchemy#10374] [Usage Questions]
Hello,  I'm using the ORM declarative model feature to map my objects to tables with several JSON columns.  My code works great in PostgreSQL database (local development), however, when I use my code to connect to the Oracle database (production environment), it gives me this error when fetching data from tables that have JSON columns:

> File "......../lib/python3.10/site-packages/sqlalchemy/sql/sqltypes.py", line 2730, in result_processor
    json_deserializer = dialect._json_deserializer or json.loads
AttributeError: 'OracleDialect_oracledb' object has no attribute '_json_deserializer'

I am using the latest python-oracledb to connect to Oracle 18c database.  My suspicion is oracledb Oracle dialect does not support JSON data type.  Any help is appreciated.

Thank you.

**Comment by zzzeek** [comment]
looks like Oracle has added a JSON datatype recently, which we do not yet have direct support for, so I've added #10375 to track that.

try creating your oracle engine like this:

```py
import json
engine = create_engine(
        "oracle://...",
)
engine.dialect._json_serializer=lambda obj: json.dumps(obj, ensure_ascii=False)
engine.dialect._json_deserializer=lambda obj: json.loads(obj)
```

see if that's all the datatype needs in your case

**Reply by TanawanK** [reply]
That works great!  Thank you very much

**Reply by zzzeek** [reply]
OK, it likely will not work for things like indexed access to the column but for insert/select should be fine

**Reply by TanawanK** [reply]
yep.. but my usage is mainly for insert and select operations only, so this is not a bad solution.

**Reply by CaselIT** [reply]
(I've unmarked as answer since in the future that will no longer be correct)

**Reply by ColeDrain** [reply]
is there any update on the support for json support on oracle?

**Reply by CaselIT** [reply]
The issue is still open, so no.
You can follow updates there

**Comment by streamnsight** [comment]
JSON type can be defined as VARCHAR2, BLOB or CLOB type with a constraint IS JSON on it.
So, when it comes to creating the tables / columns, it's possible today just setting the right type, and the constraint in the __table_args__ on the model.

```
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy import (
    BLOB,
    CheckConstraint
)

class Base(DeclarativeBase):
    pass

class MyModel(Base):
    __tablename__ = "my_models"
    __table_args__ = (
        CheckConstraint("my_json_col IS JSON"),
    )
    my_json_col: Mapped[BLOB] = mapped_column(BLOB)
```

For querying, though, it is another story. You'd need to use rawsql query at this time.
